src/language/class-modifiers.md

```diff
@@ -0,0 +1,240 @@
+---
+title: Class modifiers
+description: Modifier keywords for class declarations to control external library access.
+---
+
+Class modifiers control how a class or a mixin can be implemented, extended,
+or mixed in. 
+
+The full set of modifiers that can appear before a class declaration are `base`,
+`final`, `interface`, `sealed`, `abstract`, and `mixin`. 
+Only the `base` modifier can appear before a mixin declaration. The modifiers do
+not apply to other declarations like `enum`, `typedef`, or `extension`.
+
+When deciding whether or not to use class modifiers, consider the following questions:
+1. Is [any access](#no-modifier) fine?
+2. If not, what do you want to restrict?
+    - [Implementation](#base)?
+    - [Inheritance](#interface)?
+    - [Both](#final)?
+    - [Subtyping](#sealed) by outside libraries at all?
```

Comment:
This feels a little off.  The reason for using sealed is not that you want to prevent subtyping by outside libraries (that's what final is for).  The reason is that you want to enable exhaustiveness.

Reply:
Agree.
Rather than focus on the restrictions, I'd focus on intended use-cases and promises (to yourself or others).

So, when deciding whether to use class modifiers, consider the intended uses of the class, and which behavior the class needs to be able to rely on.


If the class is just an interface, and has only abstract members, it should probably be marked `abstract` and `interface`.
Examples are classes like `Sink` or `Future`, which are pure interfaces that others are free to implement.

(The `interface` modifier *can* be used to prevent access to an actual implementation, and force others to implement it from scratch, but there are other ways to achieve that too, so it's will not be a common use-case.)

With `base` you *ensure* that all subclasses inherit the implementation of this class. That allows you to safely call private members on every object of that type, something which allowing `implements` makes unsafe. To ensure that, no subclass is allowed to be implemented, and all subclasses must themselves be at least marked `base` to ensure that.

You should use `base` when you _need_ to rely subclasses inheriting behavior. You shouldn't add `base` just because extending is the *intended* use, and there is no good reason to implement them, because `base` forces subclasses to also be `base`, and subclasses may not want to prevent implementation.

Example of `base` classes are rare.

The `final` prevents subclassing entirely, allowing the author to safely add incremental changes to the API, and to call instance methods knowing that they haven't been overwritten in a third-party subclass.
(The `base` modifier gives most of the same guarantees, except for subclasses overriding instance members, and a `base` class with no public generative constructor could just as well have been `final`. The `final` modifier includes the effect of `base` and therefore all subclasses must again be marked at least `base`.)

The `sealed` modifier is only used to create a known (enumarable) set of subtypes, such that a switch over those types can be statically ensured to be exhaustive. It does not add any new restrictions on the subtypes of those direct subtypes. The sealed class itself cannot be instantiated.

Reply:
Thanks @lrhn, this is great. I'll be incorporating your points throughout the page, especially the angle of use cases over restrictions.